src/main/scala/edu/cmu/cs/obsidian/typecheck/Checker.scala

```diff
@@ -1468,17 +1488,19 @@ private def checkStatement(
                                     np.permission match {
                                         case Owned() =>
                                             val newType = StateType(np.contractName, Set(state._1), np.isRemote)
-                                            contextPrime.updated(x, newType).updatedMakingVariableVal(x)
-                                        case Unowned() => contextPrime
+                                            val typeFalse = StateType(np.contractName, allStates - state._1, np.isRemote)
```

Comment:
This is close, but what if this is a StateType (not just any old NonPrimitiveType)? Then we know something more specific about the possible states than just that it's one of allStates.

Reply:
@mcoblenz I put allStates - state._1, isn't that what you are referring too?

Reply:
But allStates is all the states in the contract. If the current type is a particular set of states, you should use that instead.